File: services/orders/src/commands/cancel_order.rs
use chrono::{DateTime, Local};
use uuid::Uuid;

use crate::domain::Order;

use super::{
    order_ended_producer::OrderEndedProducer,
    order_storage::{DeleteError, GetError, OrderStorage},
};

pub struct Command {
    storage: Box<dyn OrderStorage>,
    order_ended: Box<dyn OrderEndedProducer>,
}

#[derive(Debug)]
pub enum Error {
    NotFound,
    Invalid(String),
    Internal,
}

impl Command {
    pub fn new(storage: Box<dyn OrderStorage>, order_ended: Box<dyn OrderEndedProducer>) -> Self {
        Self {
            storage,
            order_ended,
        }
    }

    pub async fn cancel_order(
        &mut self,
        id: Uuid,
        canceller: Uuid,
        timestamp: DateTime<Local>,
    ) -> Result<(), Error> {
        let mut order = self.get(id).await?;
        if let Err(e) = order.cancel(canceller, timestamp) {
            return Err(Error::Invalid(e));
        }
        self.produce_event(&order).await?;
        self.delete(order.id).await
    }

    async fn get(&self, id: Uuid) -> Result<Order, Error> {
        match self.storage.get(id).await {
            Ok(o) => Ok(o),
            Err(e) => {
                log::warn!("cancel_order: {}", e);
                match e {
                    GetError::NotFound(_) => Err(Error::NotFound),
                    GetError::Unexpected(_) => Err(Error::Internal),
                }
            }
        }
    }

    async fn delete(&mut self, id: Uuid) -> Result<(), Error> {
        match self.storage.delete(id).await {
            Ok(_) => Ok(()),
            Err(e) => {
                log::warn!("cancel_order: {}", e);
                match e {
                    DeleteError::NotFound(_) => Err(Error::NotFound),
                    DeleteError::Unexpected(_) => Err(Error::Internal),
                }
            }
        }
    }

    async fn produce_event(&self, order: &Order) -> Result<(), Error> {
        match self.order_ended.produce(order).await {
            Ok(_) => Ok(()),
            Err(e) => {
                log::warn!("cancel_order: failed to produce event: {}", e);
                Err(Error::Internal)
            }
        }
    }
}

File: services/orders/commands/cancel_order.go
package commands

import (
	"context"
	"errors"
	"fmt"
	"log/slog"
	"time"

	"github.com/google/uuid"

	"github.com/indigowar/services/orders/models"
)

type CancelOrder struct {
	logger   *slog.Logger
	storage  OrderStorage
	producer OrderEndedProducer
}

func (cmd *CancelOrder) CancelOrder(
	ctx context.Context,
	orderId uuid.UUID,
	canceller uuid.UUID,
	timestamp time.Time,
) error {
	if timestamp.After(time.Now()) {
		return fmt.Errorf("cancel order: %w: timestamp", ErrInvalidRequest)
	}

	order, err := cmd.storage.Get(ctx, orderId)
	if err != nil {
		var storageErr *StorageError
		if errors.As(err, &storageErr) {
			if storageErr.ErrorType == StorageErrorTypeNotFound {
				return ErrOrderNotFound
			}
		}

		logStorageError(cmd.logger, "CancelOrder", "Get", err)
		return ErrUnexpected
	}

	if order.Customer.ID == canceller {
		return cmd.cancelOrder(ctx, order, canceller, timestamp)
	}

	if order.Restaurant == canceller {
		return cmd.cancelOrder(ctx, order, canceller, timestamp)
	}

	if order.Courier != nil {
		if order.Courier.ID == canceller {
			return cmd.cancelOrder(ctx, order, canceller, timestamp)
		}
	}

	return fmt.Errorf("cancel order: %w: canceller", ErrInvalidRequest)
}

func (cmd *CancelOrder) cancelOrder(ctx context.Context, order models.Order, canceller uuid.UUID, timestamp time.Time) error {
	order.Cancellation = &struct {
		Canceller   uuid.UUID
		CancelledAt time.Time
	}{
		Canceller:   canceller,
		CancelledAt: timestamp,
	}

	if err := cmd.producer.Produce(ctx, order); err != nil {
		return ErrUnexpected
	}

	if err := cmd.storage.Delete(ctx, order.ID); err != nil {
		logStorageError(cmd.logger, "CancelOrder", "Delete", err)
		return ErrUnexpected
	}

	return nil
}

func NewCancelOrder(logger *slog.Logger, storage OrderStorage, producer OrderEndedProducer) *CancelOrder {
	return &CancelOrder{
		logger:   logger,
		storage:  storage,
		producer: producer,
	}
}
